src/ucp/core/ucp_am.c

```diff
@@ -1132,39 +1137,68 @@ ucp_am_handler_common(ucp_worker_h worker, ucp_am_hdr_t *am_hdr, size_t hdr_size
+    int need_desc;
     ucs_status_t status;
 
-    recv_flags |= (am_flags & UCT_CB_PARAM_FLAG_DESC) ?
+    ucs_assert(total_length >= am_hdr->header_length + hdr_size);
+
+    need_desc   = !(am_flags & UCT_CB_PARAM_FLAG_DESC) &&
+                  (am_cb->flags & UCP_AM_FLAG_PERSISTENT_DATA);
+
+    recv_flags |= ((am_flags & UCT_CB_PARAM_FLAG_DESC) || need_desc) ?
                   UCP_AM_RECV_ATTR_FLAG_DATA : 0;
 
-    status      = ucp_am_invoke_cb(worker, am_hdr, hdr_size,
-                                   total_length - hdr_size, reply_ep,
-                                   recv_flags);
+    if (need_desc) {
+        /* If user requested persistent data pointer and UCT_CB_PARAM_FLAG_DESC
+         * is not set by UCT, need to initialize rdesc before calling the callback.
+         */
+        status = ucp_recv_desc_init(worker, data,
+                                    total_length - hdr_size - am_hdr->header_length,
+                                    0,
+                                    0, /* pass as a const */
+                                    0, 0, -hdr_size, &desc);
+        if (ucs_unlikely(UCS_STATUS_IS_ERR(status))) {
+            goto err;
+        }
+        data = desc + 1;
+    }
+    status = ucp_am_invoke_cb(worker, am_hdr, hdr_size, data,
+                              total_length - hdr_size, reply_ep, recv_flags);
+    if (need_desc) {
+        if (status != UCS_INPROGRESS) {
+            /* data is not needed - release previously initialized desc */
+            ucs_assert(desc != NULL);
+            ucp_recv_desc_release(desc);
+        }
+        return UCS_OK; /* can return UCS_OK only if no UCT_CB_PARAM_FLAG_DESC set */
+    }
+
     if (status != UCS_INPROGRESS) {
-        return UCS_OK; /* we do not need UCT desc, just return UCS_OK */
+        return UCS_OK; /* User does not want to keep the data, just return UCS_OK */
     }
 
     if (ucs_unlikely(!(am_flags & UCT_CB_PARAM_FLAG_DESC))) {
         ucs_error("can't hold data, FLAG_DATA flag is not set");
         return UCS_OK;
     }
 
-    ucs_assert(total_length >= am_hdr->header_length + hdr_size);
-    data   = UCS_PTR_BYTE_OFFSET(am_hdr, hdr_size + am_hdr->header_length);
     status = ucp_recv_desc_init(worker, data,
                                 total_length - hdr_size - am_hdr->header_length,
                                 0,
                                 UCT_CB_PARAM_FLAG_DESC, /* pass as a const */
                                 0, 0, -hdr_size, &desc);
     if (ucs_unlikely(UCS_STATUS_IS_ERR(status))) {
-        ucs_error("worker %p could not allocate descriptor for active"
-                  " message on callback : %u", worker, am_hdr->am_id);
-        return UCS_OK;
+        goto err;
     }
-    ucs_assert(desc != NULL);
 
     return UCS_INPROGRESS;
+
+err:
+    ucs_error("worker %p could not allocate descriptor for active"
+              " message on callback : %u", worker, am_hdr->am_id);
+    return UCS_OK;
```

Comment:
Why it is ucs_okay ? is there some retry protocol ?

Reply:
because this is UCT AM callback registered in UCT. Here we can return only 2 different statuses:
1. UCS_INPROGRESS, if we want to use data outside of this callback
2. UCS_OK , is the data is not needed and can be release by UCT